src/WebJobs.Script.WebHost/App_Start/WebHostResolver.cs

```diff
@@ -81,85 +70,81 @@ public ISecretManager GetSecretManager(WebHostSettings settings)
 
         public WebScriptHostManager GetWebScriptHostManager(WebHostSettings settings)
         {
-            if (_activeHostManager != null)
-            {
-                return _activeHostManager;
-            }
-
-            lock (_syncLock)
-            {
-                EnsureInitialized(settings);
-
-                return _activeHostManager ?? _standbyHostManager;
-            }
+            return GetActiveInstance(settings, ref _activeHostManager, ref _standbyHostManager);
         }
 
         public WebHookReceiverManager GetWebHookReceiverManager(WebHostSettings settings)
         {
-            if (_activeReceiverManager != null)
-            {
-                return _activeReceiverManager;
-            }
-
-            lock (_syncLock)
-            {
-                EnsureInitialized(settings);
-
-                return _activeReceiverManager ?? _standbyReceiverManager;
-            }
+            return GetActiveInstance(settings, ref _activeReceiverManager, ref _standbyReceiverManager);
         }
 
+        /// <summary>
+        /// This method ensures that all services managed by this class are initialized
+        /// correctly taking into account specialization state transitions.
+        /// </summary>
         internal void EnsureInitialized(WebHostSettings settings)
         {
-            if (!WebScriptHostManager.InStandbyMode)
+            lock (_syncLock)
```

Comment:
The majority of the shuffling here is due to me wrapping this all in the lock. We had the lock sprinkled all over previously, now centralizing it here. The lock is also needed for the initialization call done by the new specialization, to prevent races between incoming scale pings and the background timer.